core/storage/redis_set_store.go

```diff
@@ -81,11 +81,20 @@ func (s *RedisSetStore) List(selector string, options *ListOptions) (map[string]
 	if !strings.HasPrefix(selector, s.prefix) {
 		selector = s.prefix + selector
 	}
-	keys, err := s.client.Keys(selector).Result()
-	if err != nil {
-		return nil, err
+	var allKeys []string
+	var cursor uint64
+	for {
```

Comment:
I think you need to lock `RedisSetStore` (and `RedisMapStore`). Otherwise interleaved iterations in this `for` loop could result in both invocations only returning half of the `Scan`ned results.